From b7ce219362544d70d24d26777eb10f42c5a5efd1 Mon Sep 17 00:00:00 2001
From: Nanyong Sun <sunnanyong@huawei.com>
Date: Fri, 19 May 2023 15:38:50 +0800
Subject: [PATCH 627/780] [Huawei] memcg: support ksm merge any mode per cgroup

Offering: HULK
hulk inclusion
category: feature
bugzilla: 188921

----------------------------------------------------------------------

Add control file "memory.ksm" to enable ksm per cgroup.
Echo to 1 will set all tasks currently in the cgroup to ksm merge
any mode, which means ksm gets enabled for all vma's of a process.
Meanwhile echo to 0 will disable ksm for them and unmerge the
merged pages.
Cat the file will show the above state and ksm related profits
of this cgroup.
Signed-off-by: Nanyong Sun <sunnanyong@huawei.com>
Signed-off-by: Longlong Xia <xialonglong1@huawei.com>
---
 .../admin-guide/cgroup-v1/memory.rst          |   1 +
 include/linux/memcontrol.h                    |   4 +
 mm/memcontrol.c                               | 122 +++++++++++++++++-
 3 files changed, 124 insertions(+), 3 deletions(-)

diff --git a/Documentation/admin-guide/cgroup-v1/memory.rst b/Documentation/admin-guide/cgroup-v1/memory.rst
index c2dd6b1b3a9f..01b846ddd1f6 100644
--- a/Documentation/admin-guide/cgroup-v1/memory.rst
+++ b/Documentation/admin-guide/cgroup-v1/memory.rst
@@ -101,6 +101,7 @@ Brief summary of control files.
  memory.kmem.tcp.failcnt             show the number of tcp buf memory usage
 				     hits limits
  memory.kmem.tcp.max_usage_in_bytes  show max tcp buf memory usage recorded
+ memory.ksm                          set/show ksm merge any mode
 ==================================== ==========================================
 
 1. History
diff --git a/include/linux/memcontrol.h b/include/linux/memcontrol.h
index 372f6ec66310..0b74c38acc9b 100644
--- a/include/linux/memcontrol.h
+++ b/include/linux/memcontrol.h
@@ -396,7 +396,11 @@ struct mem_cgroup {
 #else
 	KABI_RESERVE(5)
 #endif
+#ifdef CONFIG_KSM
+	KABI_USE(6, bool ksm_merge_any)
+#else
 	KABI_RESERVE(6)
+#endif
 	KABI_RESERVE(7)
 	KABI_RESERVE(8)
 
diff --git a/mm/memcontrol.c b/mm/memcontrol.c
index 28d0ac5ced6c..a15db6a916ce 100644
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -71,6 +71,9 @@
 #include <linux/uaccess.h>
 
 #include <trace/events/vmscan.h>
+#ifndef __GENKSYMS__
+#include <linux/ksm.h>
+#endif
 
 struct cgroup_subsys memory_cgrp_subsys __read_mostly;
 EXPORT_SYMBOL(memory_cgrp_subsys);
@@ -246,10 +249,15 @@ enum res_type {
 	     iter != NULL;				\
 	     iter = mem_cgroup_iter(NULL, iter, NULL))
 
+static inline bool __task_is_dying(struct task_struct *task)
+{
+	return tsk_is_oom_victim(task) || fatal_signal_pending(task) ||
+		(task->flags & PF_EXITING);
+}
+
 static inline bool task_is_dying(void)
 {
-	return tsk_is_oom_victim(current) || fatal_signal_pending(current) ||
-		(current->flags & PF_EXITING);
+	return __task_is_dying(current);
 }
 
 /* Some nice accessors for the vmpressure. */
@@ -5259,6 +5267,104 @@ static ssize_t memory_reclaim(struct kernfs_open_file *of, char *buf,
 	return nbytes;
 }
 
+#ifdef CONFIG_KSM
+static int memcg_set_ksm_for_tasks(struct mem_cgroup *memcg, bool enable)
+{
+	struct task_struct *task;
+	struct mm_struct *mm;
+	struct css_task_iter it;
+	int ret = 0;
+
+	if (enable == READ_ONCE(memcg->ksm_merge_any))
+		return 0;
+
+	css_task_iter_start(&memcg->css, CSS_TASK_ITER_PROCS, &it);
+	while (!ret && (task = css_task_iter_next(&it))) {
+		if (__task_is_dying(task))
+			continue;
+
+		mm = get_task_mm(task);
+		if (!mm)
+			continue;
+
+		if (mmap_write_lock_killable(mm)) {
+			mmput(mm);
+			continue;
+		}
+
+		if (enable)
+			ret = ksm_enable_merge_any(mm);
+		else
+			ret = ksm_disable_merge_any(mm);
+
+		mmap_write_unlock(mm);
+		mmput(mm);
+	}
+	css_task_iter_end(&it);
+
+	return ret;
+}
+
+static int memory_ksm_show(struct seq_file *m, void *v)
+{
+	unsigned long ksm_merging_pages = 0;
+	unsigned long ksm_rmap_items = 0;
+	long ksm_process_profits = 0;
+	unsigned int tasks = 0;
+	struct task_struct *task;
+	struct mm_struct *mm;
+	struct css_task_iter it;
+	struct mem_cgroup *memcg = mem_cgroup_from_seq(m);
+
+	css_task_iter_start(&memcg->css, CSS_TASK_ITER_PROCS, &it);
+	while ((task = css_task_iter_next(&it))) {
+		mm = get_task_mm(task);
+		if (!mm)
+			continue;
+
+		if (test_bit(MMF_VM_MERGE_ANY, &mm->flags))
+			tasks++;
+
+		ksm_rmap_items += mm->ksm_rmap_items;
+		ksm_merging_pages += mm->ksm_merging_pages;
+		ksm_process_profits += ksm_process_profit(mm);
+		mmput(mm);
+	}
+	css_task_iter_end(&it);
+
+	seq_printf(m, "merge any state: %d\n", READ_ONCE(memcg->ksm_merge_any));
+	seq_printf(m, "merge any tasks: %u\n", tasks);
+	seq_printf(m, "ksm_rmap_items %lu\n", ksm_rmap_items);
+	seq_printf(m, "ksm_merging_pages %lu\n", ksm_merging_pages);
+	seq_printf(m, "ksm_process_profits %ld\n", ksm_process_profits);
+	return 0;
+}
+
+static ssize_t memory_ksm_write(struct kernfs_open_file *of, char *buf,
+					size_t nbytes, loff_t off)
+{
+	bool enable;
+	int err;
+	struct mem_cgroup *memcg = mem_cgroup_from_css(of_css(of));
+
+	buf = strstrip(buf);
+	if (!buf)
+		return -EINVAL;
+
+	err = kstrtobool(buf, &enable);
+	if (err)
+		return err;
+
+	err = memcg_set_ksm_for_tasks(memcg, enable);
+	if (err)
+		return err;
+
+	WRITE_ONCE(memcg->ksm_merge_any, enable);
+
+	return nbytes;
+}
+#endif /* CONFIG_KSM */
+
 #ifdef CONFIG_CGROUP_V1_WRITEBACK
 #include "../kernel/cgroup/cgroup-internal.h"
 
@@ -5577,6 +5683,14 @@ static struct cftype mem_cgroup_legacy_files[] = {
 		.seq_show = wb_blkio_show,
 		.write = wb_blkio_write,
 	},
+#endif
+#ifdef CONFIG_KSM
+	{
+		.name = "ksm",
+		.flags = CFTYPE_NOT_ON_ROOT,
+		.write = memory_ksm_write,
+		.seq_show = memory_ksm_show,
+	},
 #endif
 	{ },	/* terminate */
 };
@@ -5822,7 +5936,9 @@ mem_cgroup_css_alloc(struct cgroup_subsys_state *parent_css)
 		if (parent != root_mem_cgroup)
 			memory_cgrp_subsys.broken_hierarchy = true;
 	}
-
+#ifdef CONFIG_KSM
+	memcg->ksm_merge_any = false;
+#endif
 	/* The following stuff does not apply to the root */
 	if (!parent) {
 		root_mem_cgroup = memcg;
-- 
2.17.1

